firecracker-control/local.go

```diff
@@ -77,20 +80,38 @@ func newLocal(ic *plugin.InitContext) (*local, error) {
 	return &local{
 		vm:          make(map[string]*instance),
 		rootPath:    ic.Root,
+		publisher:   ic.Events,
 		findVsockFn: findNextAvailableVsockCID,
 	}, nil
 }
 
 // CreateVM creates new Firecracker VM instance
-func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*proto.CreateVMResponse, error) {
-	id, err := s.makeID()
+func (s *local) CreateVM(ctx context.Context, req *proto.CreateVMRequest) (*empty.Empty, error) {
+	var (
+		id     = req.GetVMID()
+		logger = log.G(ctx).WithField("vm_id", id)
+	)
+
+	if id == "" {
+		return nil, errors.New("invalid VM ID")
+	}
+
+	s.vmLock.Lock()
```

Comment:
Doesn't have to be addressed now, but it is a bit unfortunate that we aren't spinning up VMs in parallel anymore (though it remains to be seen if that is a performance bottleneck).

I think we could still spin up in parallel if we actually add an entry to `s.vm` for the ID right at the beginning of the method and just have a defer that removes the entry in case an error is encountered (each individual operation being wrapped in a quick lock/unlock).

I'm not convinced that complication is necessary at this stage, so no strong feelings either way for this PR.

Reply:
Yep. Will add to TODO list.